### ISSUE TITLE
Flush the log file

### ISSUE THREAD

**Issue by mpmxyz** [Yellow-Dog-Man/Resonite-Issues#508]
### Is your feature request related to a problem? Please describe.

The stream that is used to write the log file is not flushed after every line.
This means that parts of the log may not be written to file for minutes or longer.
So the log files uploaded in bug reports may sometimes not contain the described bug or just parts of it.
PS: I know that the instructions are "DO NOT USE MODS. Start the client, replicate the issue and **exit immediately**." but I didn't see that part for the longest time and I can guess there are many more who are not even aware that this kind of problem exists.

### Describe the solution you'd like

The log file should be flushed regularly/after every write to make finding bugs more reliable.

### Describe alternatives you've considered

Raising more awareness about the technical limitations of writing to log files may be a possibility but I see two disadvantages:
1. There will still be users who will not be aware and may unknowingly upload incomplete log files.
2. Modders will have a harder time debugging a mod at runtime if they can't rely on logfiles being up to date.

### Additional Context

[USER - 2023.10.20.831 - 2023-11-02 23_37_50 - Snapshot.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13245401/USER.-.2023.10.20.831.-.2023-11-02.23_37_50.-.Snapshot.log)
a log file that is not yet completely written - stayed like that for >5 minutes (personal data removed)

**Comment by Nytra** [comment]
I think if you exit Resonite it will write everything as normal. However it is annoying that this happens while the app is running.

A little workaround for this is if you spam the F3 key it will write some debug texts to the log which will make it update.

**Comment by Nytra** [comment]
I made a mod for this https://github.com/Nytra/ResoniteLogFlush

**Comment by Frooxius** [comment]
This is intentional. Flushing requires IO operations which have a performance overhead. Flushing after every line would have a performance impact. 

We request users to shut down Resonite before sending the log, which I think is perfectly reasonable request, so this is only a problem if user breaks that issue. 

If you want this behavior for mods, then handle the flushing in the mod. We will not introduce this officially.

**Comment by mpmxyz** [comment]
I found a new perspective on this request we didn't talk about in this issue: Crashes
You can't manually shut down the engine after them.
I now have log files with up to 17 seconds of log information missing.
This means I lack information of the events immediately prior to the crash. (e.g. a user joining or a handled exception that may have corrupted the engine state to cause the crash)

[Crash_2024-03-23_213152739.zip](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/14735819/Crash_2024-03-23_213152739.zip) (Big disclaimer: This specific instance is _tainted by modding_!)

It was a series of crashes within a short timeframe. Unfortunately a good amount of users left and possibly problematic items have been removed before I switched off mods so I couldn't replicate the crash a fourth time.

I've taken this as a reason to look into the performance impact of flushing on my local machine.
For this I inserted the following test code via modding:
```C#
//flush before test -> start clean
FrooxEngineBootstrap.LogStream.Flush();
var results = new List<(bool, int, double)>();
var sw = new Stopwatch();
foreach (bool flushed in new bool[] { false, true })
{
    for (int i = 0; i < 100; i++)
    {
        sw.Restart();
        FrooxEngineBootstrap.LogStream.WriteLine("Log Performance Test #"+i);
        //UniLog.Log("Test message", false);
        //UniLog.Log("Test message", true);
        if (flushed)
        {
            FrooxEngineBootstrap.LogStream.Flush();
        }
        sw.Stop();
        results.Add((flushed, i, sw.GetElapsedMilliseconds()));
    }
}
UniLog.Log("Flushed  i   time(ms)");
foreach (var (flushed, i, time) in results)
{
    UniLog.Log($"{flushed}  {i} {time}");
}
```
I post-processed the test results to more easily read csv files:
[logging via FrooxEngineBootstrap.LogStream.WriteLine](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/14735979/FrooxEngineBootstrap.LogStream.WriteLine.csv): flush overhead = 0.35ms/100 calls (+609%)
[logging via UniLog.Log without stacktrace](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/14735962/UniLog.Log.no.stacktrace.csv): flush overhead = 3.9ms/100 calls (+342%)
[logging via UniLog.Log with stacktrace](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/14735968/UniLog.Log.with.stacktrace.csv): flush overhead = 13ms/100 calls (+16%)

In conclusion I'd say that there is a small but at least measurable impact. (worst case: 2% fps-loss per log line per frame with stack traces enabled)
A good middle-ground would be to make it an option. I personally would have it enabled and be OK with the loss as there is rarely more than a single log entry per frame.

PS: Enabling stacktraces is a much worse offender, with an overhead of 81ms/100 calls. (+7226%)

**Comment by mpmxyz** [comment]
I'd like to add another point pro-flushing by using an example:
I was helping a user today where [Resonite was stuck frozen within the loading screen](https://discord.com/channels/1040316820650991766/1260919213737115810).
The log file was cut off just before the database should have been loaded:
[YELNA - 2024.7.10.1399 - 2024-07-11 07_43_37.log](https://github.com/user-attachments/files/16181215/YELNA.-.2024.7.10.1399.-.2024-07-11.07_43_37.log)
With my experience of reverse-engineering Resonite, a very detailed back and forth between the user and me and around 1.5 hours of time we finally discovered that failing to start SystemHelperServer.exe was halting Resonite.
My 1.5 hours of time weren't the first. Someone else was also trying but could not figure anything out.
This wouldn't have happened if a background thread ensured the log would be flushed every 10 seconds or so. And it wouldn't have as much of an impact as flushing every line would have.
